[FSDP][8/N] Replace `_FSDPPolicy.policy` with `_Policy._run_policy`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104986
* __->__ #104969
* #104999
* #104967
* #104427

This does some code organization improvement. 
- It renames `_FSDPPolicy` to `_Policy` to show that it is not only for FSDP but for any module-level API.
- It formalizes the contract that such a policy should return something like `target_module_to_kwargs: Dict[nn.Module, Dict[str, Any]]` that maps each module to wrap to its kwargs. It does so by requiring a `_run_policy` abstract method (this time private since users do not need to care about it). Then, our auto wrapping can just call `_run_policy()` to generate the dict and do any validation or post-processing.

This PR is technically BC-breaking because it removes the public `ModuleWrapPolicy.policy`. However, I do not think anyone was using that anyway, so this is a pretty safe breakage.

Commit:
[FSDP][8/N] Replace `_FSDPPolicy.policy` with `run_policy`

[ghstack-poisoned]

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `run_policy`"




[ghstack-poisoned]

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `run_policy`"




[ghstack-poisoned]

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `run_policy`"




[ghstack-poisoned]

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `_run_policy`"


This is a code organization improvement. It formalizes the contract that an FSDP policy should return something like `target_module_to_kwargs: Dict[nn.Module, Dict[str, Any]]` that maps each module to wrap to its FSDP kwargs. It does so by requiring a `_run_policy` abstract method (this time private since users do not need to care about it). Then, our auto wrapping can just call `_run_policy()` to generate the dict and do any validation or post-processing.

This PR is technically BC-breaking because it removes the public `ModuleWrapPolicy.policy`. However, I do not think anyone was using that anyway, so this is a pretty safe breakage.



[ghstack-poisoned]

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `_run_policy`"


This is a code organization improvement. 
- It renames `_FSDPPolicy` to `_WrapPolicy` to show that it is not only for FSDP but for any module-level API.
- It formalizes the contract that such a policy should return something like `target_module_to_kwargs: Dict[nn.Module, Dict[str, Any]]` that maps each module to wrap to its kwargs. It does so by requiring a `_run_policy` abstract method (this time private since users do not need to care about it). Then, our auto wrapping can just call `_run_policy()` to generate the dict and do any validation or post-processing.

This PR is technically BC-breaking because it removes the public `ModuleWrapPolicy.policy`. However, I do not think anyone was using that anyway, so this is a pretty safe breakage.



[ghstack-poisoned]

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `_run_policy`"


This does some code organization improvement. 
- It renames `_FSDPPolicy` to `_WrapPolicy` to show that it is not only for FSDP but for any module-level API.
- It formalizes the contract that such a policy should return something like `target_module_to_kwargs: Dict[nn.Module, Dict[str, Any]]` that maps each module to wrap to its kwargs. It does so by requiring a `_run_policy` abstract method (this time private since users do not need to care about it). Then, our auto wrapping can just call `_run_policy()` to generate the dict and do any validation or post-processing.

This PR is technically BC-breaking because it removes the public `ModuleWrapPolicy.policy`. However, I do not think anyone was using that anyway, so this is a pretty safe breakage.



[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/wrap.py:
```diff
@@ -195,21 +178,32 @@ def _module_wrap_policy(
+                continue
+            elif isinstance(module, module_classes):
+                # Shallow copy to avoid coupling changes across modules
+                target_module_to_kwargs[module] = copy.copy(root_kwargs)
```
why do we need to copy the args? Do we expect root_kwargs to change?

awgu replied:
We must copy the kwargs or else if a function overrides it (e.g. the mixed precision overrides), then it will override it for _every_ target module. We have to avoid them sharing a reference to the same kwarg dict.

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `_Policy._run_policy`"


This does some code organization improvement. 
- It renames `_FSDPPolicy` to `_Policy` to show that it is not only for FSDP but for any module-level API.
- It formalizes the contract that such a policy should return something like `target_module_to_kwargs: Dict[nn.Module, Dict[str, Any]]` that maps each module to wrap to its kwargs. It does so by requiring a `_run_policy` abstract method (this time private since users do not need to care about it). Then, our auto wrapping can just call `_run_policy()` to generate the dict and do any validation or post-processing.

This PR is technically BC-breaking because it removes the public `ModuleWrapPolicy.policy`. However, I do not think anyone was using that anyway, so this is a pretty safe breakage.



[ghstack-poisoned]

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `_Policy._run_policy`"


This does some code organization improvement. 
- It renames `_FSDPPolicy` to `_Policy` to show that it is not only for FSDP but for any module-level API.
- It formalizes the contract that such a policy should return something like `target_module_to_kwargs: Dict[nn.Module, Dict[str, Any]]` that maps each module to wrap to its kwargs. It does so by requiring a `_run_policy` abstract method (this time private since users do not need to care about it). Then, our auto wrapping can just call `_run_policy()` to generate the dict and do any validation or post-processing.

This PR is technically BC-breaking because it removes the public `ModuleWrapPolicy.policy`. However, I do not think anyone was using that anyway, so this is a pretty safe breakage.



[ghstack-poisoned]

Commit:
Update on "[FSDP][8/N] Replace `_FSDPPolicy.policy` with `_Policy._run_policy`"


This does some code organization improvement. 
- It renames `_FSDPPolicy` to `_Policy` to show that it is not only for FSDP but for any module-level API.
- It formalizes the contract that such a policy should return something like `target_module_to_kwargs: Dict[nn.Module, Dict[str, Any]]` that maps each module to wrap to its kwargs. It does so by requiring a `_run_policy` abstract method (this time private since users do not need to care about it). Then, our auto wrapping can just call `_run_policy()` to generate the dict and do any validation or post-processing.

This PR is technically BC-breaking because it removes the public `ModuleWrapPolicy.policy`. However, I do not think anyone was using that anyway, so this is a pretty safe breakage.



[ghstack-poisoned]

awgu commented:
@pytorchbot merge -f "unrelated failure"
